### PR TITLE
k6n10f DSP behavioral model

### PR DESCRIPTION
**Pull request by mkurc-ant**
This PR adds behavioral model of type 1 DSP. Based on https://github.com/SymbiFlow/yosys-symbiflow-plugins/pull/226 which has to be merged first.

This is still work in progress.

### CODE REVIEW

File: ql-qlf-plugin/qlf_k6n10f/cells_sim.v [hunk 1]
```diff
@@ -664,22 +664,249 @@ module TDP_BRAM36 (
...
+)(
+    input  [NBITS_A-1:0] a_i,
+    input  [NBITS_B-1:0] b_i,
+    output [NBITS_Z-1:0] z_o,
+    output [NBITS_B-1:0] dly_b_o,
+
+    input         unsigned_a_i,
+    input         unsigned_b_i,
+
+    input         clock_i,
```
[the remark this answers is not in the document]
Comment (mkurc-ant) [reply ql-qlf-plugin/qlf_k6n10f/cells_sim.v:708]: Right, I'll fix that in https://github.com/SymbiFlow/yosys-symbiflow-plugins/pull/226 as this one is based on that

File: ql-qlf-plugin/qlf_k6n10f/dsp_map.v [hunk 2]
```diff
@@ -6,22 +6,77 @@
...
+        .feedback_i         (2'b00),
+        .round_i            (1'b0),
+        .shift_right_i      (1'b0),
+        .saturate_enable_i  (1'b0),
+        .output_select_i    (1'b0)
+    );
+
+    assign Y = z[37:0];
+
 endmodule
```
[the remark this answers is not in the document]
Comment (mkurc-ant) [reply ql-qlf-plugin/qlf_k6n10f/dsp_map.v:44]: I haven't considered any mode bit mapping yet. Basically the operating mode is defined by connecting the control ports, for now they are all tied to 0. You may look at https://github.com/SymbiFlow/yosys-symbiflow-plugins/pull/226 where I described the multiplier techmapping process in more detail.

**Comment by rakeshm75** [comment]
There are few observation with respect to the DSP block instantiation in dsp_map.v file.

There are few port mismatch:
1. inputs unsigned_a_i & unsigned_b_i on the dsp_t1_xxx is not connected. I see that signed and unsigned logic control is in the dsp_map.v file but that logic is already part of the DSP IP block
2. Input load_acc_i on the dsp_t1_xxx is not connected
3. output_select_i is 3 bit vector signal ([2:0]) but in dsp_map.v its a single bit scalar signal 
4. shift_right_iis 6 bit vector signal ([5:0]) but in dsp_map.v its a single bit scalar signal
5. parameters NBITS_ACC, NBITS_A, NBITS_B & NBITS_Z need to be passed into dsp_t1_xxx
6. what happens to clock and reset signals?

**Comment by lpawelcz** [comment]
@tpagarani DSPs model were updated according to the most recent examples. I believe it is ready for final review.

**Comment by tpagarani** [comment]
@rakeshm75 please help review

**Comment by rakeshm75** [comment]
Here are few observations:

Currently, I see that the DSP model (dsp_t1_sim) is under dsp_t1_20x18x64 & dsp_t1_10x9x32.

Ideally the dsp model should be under the QL_DSP2 module:

module QL_DSP2
              .
              .
              .
          dsp_t1_sim (
                     .
                     .
                     );
endmodule

And QL_DSP2 should be port mapped under dsp_t1_20x18x64 & dsp_t1_10x9x32:

module dsp_t1_20x18x64 
              .
              .
              .
          QL_DSP2 (
                     .
                     .
                     );
endmodule

module dsp_t1_10x9x32
              .
              .
              .
          QL_DSP2 (
                     .
                     .
                     );
endmodule

**Comment by rakeshm75** [comment]
In the dsp model (dsp_t1_sim) the shift_d2 assignment is wrong:

On line 825 current code is as follows:
                  wire [5:0] shift_d2 = output_select_i[1] ? r_shift_d2 : shift_right_i; 
it should be:
                  wire [5:0] shift_d2 = output_select_i[1] ? shift_d1 : r_shift_d2;

**Comment by lpawelcz** [comment]
@rakeshm75 thank you for your comments. I changed the structure of DSP modules as you suggested, updated `shift_d2` assignment and rebased the code onto newest master resolving the conflicts.

**Comment by rakeshm75** [comment]
Looks good to me, approved.

**Comment by mkurc-ant** [comment]
Ok, merging.

**Pull request merged**